doc: Add user guide for memory footprint optimization.

grochu opened the pull request:
A doc-only improvement related to NCSIDB-228. It provides a short description on how to optimize memory footprint. It should help develop code for the smallest nRF devices e.g. nRF52810, 811. It is largely based on: https://github.com/nrfconnect/sdk-nrf/pull/3157.

joerchan commented on doc/nrf/ug_app_dev.rst:
```diff
@@ -56,3 +56,46 @@ In order to define your own board, you can use the following Zephyr guides as re
+  * GATT Service Changed, etc.
+
+* Reduce the stack sizes of the Bluetooth internal threads where possible (based on the output of the Thread Analyzer):
+ * :option:`ONFIG_SDC_RX_STACK_SIZE`,
```
```suggestion
 * :option:`CONFIG_SDC_RX_STACK_SIZE`,
```

grochu replied:
fixed

joerchan commented on doc/nrf/ug_app_dev.rst:
```diff
@@ -56,3 +56,46 @@ In order to define your own board, you can use the following Zephyr guides as re
+When developing an application, ROM and RAM footprint are important factors, especially when the firmware runs on the most resource-constrained devices e.g. nRF52810 or nRF52811.
+If you want your application to use the minimum required resources, follow the steps below and tune the |NCS| configuration parameters reducing the memory footprint:
+
+# Follow the basic hints for Zephyr :ref:`zephyr:optimizations`
```
We could also point to this sample in upstream: https://docs.zephyrproject.org/2.3.0/samples/basic/minimal/

grochu replied:
Good point - added

joerchan commented on doc/nrf/ug_app_dev.rst:
```diff
@@ -56,3 +56,46 @@ In order to define your own board, you can use the following Zephyr guides as re
+  * GATT Caching,
+  * GATT Service Changed, etc.
+
+* Reduce the stack sizes of the Bluetooth internal threads where possible (based on the output of the Thread Analyzer):
```
We could refer to the documentation of the thread analyzer here:
https://docs.zephyrproject.org/2.3.0/guides/debugging/thread-analyzer.html

grochu replied:
It is linked in line 67, but I added a link here too.

greg-fer commented:
This is too significant a change to docs to squeeze it in for 1.4.0. Let's review it calmly and implement after 1.4.0 release.
Removing milestone 1.4.0.

ru-fu reviewed (changes requested):
I'll add a commit with updates.

ru-fu commented:
I added a commit with language fixes - please check and squash, if okay.

Can we add information for more subsystems?

grochu commented:
> I added a commit with language fixes - please check and squash, if okay.
> 
> Can we add information for more subsystems?

Squashed. I modified line 96 to avoid confusion between the thread stacks (memory object) and the Bluetooth stack (protocol implementation).
For other subsystems we need more authors to contribute. Would be great to fill this section.

ru-fu commented:
> > Can we add information for more subsystems?
> 
> For other subsystems we need more authors to contribute. Would be great to fill this section.

Who can contribute? @LuDuda , @wbober, @hakonfam , @frkv  maybe?

grochu commented:
@ru-fu can we go ahead with this PR for 1.4.0 and add other sections after the release? This PR was raised to specifically help developing BLE a[[lications for the smallest devices (NCSIDB-228).

greg-fer commented:
We're not going to merge it to 1.4.0.
This is going to be available on latest in a couple of days, when finalized.
We are beyond doc freeze, and this PR still needs some work.

frkv commented:
I will provide some limited information, but this must be done after the release. This will specifically be about the nrf_security module.

There also needs to be a chapter on TF-M integration (and secure vs non-secure memory management). We are not done with the configurability in that solution yet, so the additions to theis user-guide may require two tranches

ru-fu reviewed (approved):
It's okay to go in for now, but should be expanded later.

joerchan commented:
> Following a Teams chat with Michal and Shantha, I've decided not to add it to 1.4.0.
> This is going to be available on latest in a couple of days, when finalized.
> We are beyond doc freeze, and this PR still needs some work.

Yes, I accepted that when you first removed it from the 1.4.0 milestone, probably @carlescufi didn't notice that it had been removed already.

greg-fer commented:
@ru-fu About other subsystems, should they be documented on this page or elsewhere?

ru-fu commented:
> 
> 
> @ru-fu About other subsystems, should they be documented on this page or elsewhere?

On this page for now, parallel to Bluetooth.

greg-fer commented:
> @ru-fu About other subsystems, should they be documented on this page or elsewhere?
> > On this page for now, parallel to Bluetooth.

@lmaciejonczyk , with regards to our conversation yesterday, please open a PR that adds a Thread section to this page.
